Changes to support W3C style IDs and propagation

vancem opened the pull request:
See https://w3c.github.io/trace-context for  more on what W3C Ids are.  

This is mostly for design discussion purposes.   Adding NO MERGE flag until we work through the details.   (we will need a design review as well, but first things first).   

Also see https://github.com/aspnet/Hosting/pull/1577/files for changes to the ASP.NET hosting code to support incoming HTTP propagation of the ID and TraceState.  

This implementation was meant to be pretty much the 'minimal' change from where we are today that would seem to support the W3C style IDs.   It does not (yet) support some important considerations including

1. 'Always On' capability.  
2. Perf  / Sampling.  To always be on, it has to be cheap.  
3. Versioning.  We are likely to need to tweek it multiple times over the next year or so.  Thus we need a way of allowing this without forcing everyone to update there .NET Framework.  

This work includes feedback from Noah where we try to keep things super-simple and avoid 'support both' functionality if at all possible.   Ideally in 5 years we only have one format (the W3C format), and everything else is only available via compatibility switches (or not supported).  

You can see the changes to Activity are pretty minimal.  Basically 

1. we add TraceStateString (which is really just a first class baggage item), 
2.  provide a new IDFormat property as well as a STATIC (DefaultIdFormat and ForceDefaultIdFormat) which tells Activity which format to use (Use the default when you don't have information on a parent OR ForceDefaultIdFormat is set).   
3.  Added TraceId.   This is basically a type that provides another way of getting at the RootId property when it is a W3C ID format.  This is useful because it can be more efficient than using strings (which is what RootId is).   It is a 16 byte binary ID (often expressed as a hex string).  
4. Added SpanId  This is a type that provides a way of getting at the 'rest' of the ID' that is not the TraceId (for W3C ID format).    It is a 8 byte binary ID (often expressed as a hex string).  

@noahfalk  @lmolkova  @glennc  @SergeyKanzhelev @davidfowl @pakrym

lmolkova commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +347,77 @@ public void Stop()
+        /// However if the activity has no parent, it has to do something.   
+        /// This sets it so that WC3 format is used in this case.  
+        /// </summary>
+        static public bool UseW3CFormat { get; set; }
```
it should be set to true by default.
Assuming it's false but some incoming requests have W3C headers, we'll create mixed w3c and 'legacy' activities - this would create an inconsistent experience and not going to help much.

I realize that anyone can change it to make it right, but by default let's be consistent.

vancem replied:
The default is definitely an interesting question.    My concern is compatibility.  Maybe we just break things, but normally we go through a procedure where we support both for a while first with the default the old way, then the default the new way (with a app compat config that will flip it back), and then get rid of the old way.   

I note that even if the default was off, if things come in from the outside (the most common case), with W3C headers, the code WILL do the right thing.   Thus flipping the default is that THAT important.   The question is how important is current compat.    Generally we find out that it is more important than we first think (people did build on what we shipped).  

I would like to defer this decision, until we get to a later phase of the checkin.

lmolkova commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +347,77 @@ public void Stop()
+            string newSpanId = Guid.NewGuid().ToString("n").Substring(16, 16);
+            if (ParentId != null)
+            {
+                return ParentId.SubString(0, 36) + newSpanId + ParentId.SubString(52, 3);
```
I realize this is a strawman, but just want to clarify that **we want to have access to newSpanId and parentSpanId without parsing a string**. As we want to decouple http wire format from how telemetry is reported. And we'll potentially access it many times during Activity lifetime, so it should be more efficient to keep them rather than calculate each time.

lmolkova replied:
Assuming there was Request-Id in the incoming request and SetParentId was called with |abc...1.2.3, rootId may become invalid

We should validate ParentId here and if it's not valid one, we should generate a new traceId

vancem replied:
I am interested in insure we don't have a bunch of unnecessary string maniputation, so yes, I am open to having activity cache/parse things, but I want to understand the issue a bit more first.  In particualr I want to understand why you want SpanIds AT RUNTIME.   I can see wanting them at PROCESSING TIME, but otherwise it is not clear to me (seems like you pass the ID around.   

It actually seems MORE VERSION RESILIANT, if the RUNTIME part of the system treats to the extent possible as a opaque blob (which is what Activity.Id is),   This allows the format of the ID to change (add more information) without breaking all existing code.   Thus ideally you should not be parsing the ID, and certainly not multiple times.    Can you describe when you think you need to parse the spans?

vancem replied:
> We should validate ParentId here and if it's not valid one, we should generate a new traceId

The logic behind the code is that in SetParentId, if we detect that the ID is a valid W3CId we set the 'IsW3CFormat' bit.  We only come into this code in the W3CId case.

In your scneario (where the incomming request is using the old format, we don't ever run this code (we use the old format and do the old thing).     

The expectation is that systems that wish to only support the new W3C format will test the IsW3CFormat bit and drop the activity if it is not true.

lmolkova replied:
>   I want to understand why you want SpanIds AT RUNTIME. I can see wanting them at PROCESSING TIME, but otherwise it is not clear to me (seems like you pass the ID around.

> It actually seems MORE VERSION RESILIANT, if the RUNTIME part of the system treats to the extent possible as a opaque blob (which is what Activity.Id is), This allows the format of the ID to change (add more information) without breaking all existing code. Thus ideally you should not be parsing the ID, and certainly not multiple times. Can you describe when you think you need to parse the spans?


When tracing system reports telemetry (it's applicable to AppInsights, OpenCensus, Zipkin, Jaeger, NewRelic and Dynatrace and probably the rest of them) it records traceId, parentSpanId and spanId separately. So at least at the moment of recoding telemetry item we should have them all. In reality, we have more then one telemetry reported in scope of the activity: log, exception or some event that requires traceId and spanId to be properly set on such telemetry as well for correlation.   

This ids are used for indexing and sql queries on the backends. 

Does it explain my point?

lmolkova replied:
and yes, we could actually save whole W3CId as span id, but this is redundunt as we save  traceId in the other field anyway.

lmolkova replied:
> The logic behind the code is that in SetParentId, if we detect that the ID is a valid W3CId we set the 'IsW3CFormat' bit. We only come into this code in the W3CId case.

We generate W3C Id if DefaultIdFormat. Which we are going to set tot true in all possible cases.
```csharp
                if (IsW3CId || DefaultIdFormat == ActivityIDFormat.W3C)
                {
                    IsW3CFormat = true;
                    Id = GenerateW3CId();
                }
```

vancem replied:
You found a bug.  The logic should be 
```
                if (IsW3CId || ParentId == null && DefaultIdFormat == ActivityIDFormat.W3C)
```
I have updated the PR.

lmolkova replied:
This is a feature. With && it creates even more complications because now the upstream service controls the protocol and tracing system or user cannot prevent it from happening.

let's assume service X runs on AspNetCore 2.0 with W3C Activity. 
It is called by A and B upstream services and it calls C and D  downstream services.

Let's assume X upgrades to corefx 3.0, but others are 2.0. Things are good, everything works per over old protocol.

Service X does not necessarily control how A, B, C, or D update. It's enough for any of the upstream services to update to break correlation from A to C or D: 

After that A updates to 3.0 and then requests coming from  A to X are traceable, but not further.
Request from B are traceable end-to-end, 

As we don't know customer's layout we have to provide them with a way to force W3C or force Request-Id.  Following upstream protocol does not help anyone,

My proposal is to 
- allow to force W3C ids generation in all cases (even if parent is request-id)
- emit W3C header in all cases (always or when forced) 
- emit Request-Id in format `|traceid.spanid.` (always or when enabled, or leave it to the tracing system)
- generate old ids only when request-id is enforced, change RootId algo generation to be the same as traceId, so that reasonable mapping from Activity.Id to w3c would be possible anyway

X decides to update. It forces W3C and emits Request-Id with `|traceid.spanid` value. Correlation with all downstream services works.
This does not require downstream customers to even update AspNetCore/HttpClient.

So all customers start to upgrade from the topmost service/client they own, and correlation works and dependency calls troubleshooting works in all cases. This way they control their part of infra and not controlled by parts they don't own.

Correlation with upstream services will be broken unless tracing system can support multiple parents for the same request - which we just enabled.

However depending on the layout, customers may choose to turn W3C off and update bottom-up.

vancem replied:
> When tracing system reports telemetry (it's applicable to AppInsights, OpenCensus, Zipkin, Jaeger, NewRelic and Dynatrace and probably the rest of them) it records traceId, parentSpanId and spanId separately

I understand.   What are the data types for this data?   Presumably spanIds are passed as a 64 bit long?   But the TraceId is too big for that.  What do they use in their signature to represent it?

lmolkova replied:
AI - strings
OpenCensus - byte[]
Zipkin (and perhaps Jaeger) - strings.

I can check for the rest, but let's assume it's strings

vancem replied:
Thanks for the info on how TraceId is encoded.   I am assuming at this point that logging systems don't care about the Flags part of the ID (assuming we have 'Recorded' bit on Activity).   

I need to think about your feedback on the disadvantages of following the convention of the incoming request.

vancem replied:
One other quick question.   The TraceId is a string.  Is the string just a 128bit number in Hex?   (I am assuming yes).

lmolkova replied:
it's 16 bytes, i.e. 128 bit number

vancem replied:
Right,  a 128 bit (16 byte) number would take 32 characters to express as a string if you use Hex (2 chars per byte).   Is that what they do?

lmolkova replied:
Exactly, thats what they do

lmolkova commented:
let's assume new Activity is used with old HttpClient & old Asp.Net Core which are not aware of W3C.
It's totally possible for customers that run the latest AppInsights on e.g. netcoreapp 2.0.

depending on `UseW3CFormat` Activity may generate W3C Id, which will be injected by HttpClient into the Request-Id header which will not be useful for anyone. 

If we want to be backward compatible we can try to detect such cases on tracing system level and inject traceparent ourselves, but I'd like to avoid this level of hacks. 
So I still propose to add another property with public getter: call it TraceParent or W3CId, which constructs new w3c Id.  And obsolete `Id` and make it return `|traceid.spanId.` This will allow tracing systems to glue up different protocols and enable a transition story for our customers.

HttpClient in this case will send 

```csharp
if (currentActivity.IsWC3Id)
  request.Headers.Add(DiagnosticsHandlerLoggingStrings.TraceParentHeaderName, currentActivity.W3CId);
else
  request.Headers.Add(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName, currentActivity.Id);

```

noahfalk commented:
> let's assume new Activity is used with old HttpClient & old Asp.Net Core which are not aware of W3C.

Under what situation would someone want to set Activity.UseW3CFormat = true, while at the same time continuing to use legacy software components in their services that do not support W3C?

The closest scenario I am envisioning looks like this:
a) I have a big distributed system and updating the software in it is hard to do, or not entirely under my control. Lets say it has nodes X, Y, and Z in it. Requests arrive at X, then X sends requests to Y and Z.
b) Currently this system communicates using the RequestID standard and various portions of it are old so those parts have no support for anything else.
c) But I want to make X better by sending requests to a new service W and this external service only speaks W3C
d) Its important that I can get distributed tracing across the whole thing
e) Thankfully X, Y, Z, and W all use a logging framework that is able to understand both RequestID and W3C, and it can correlate the two together as long as the IDs on both sides follow some conventions for translating.


In this case I'd still be scared to set Activity.UseW3CFormat = true in node A. In particular with HttpClient implemented as suggested that means that all my requests to Y and Z could switch from RequestID to TraceParent and by assumption Y and Z probably have old software in them that doesn't understand TraceParent. Because updating the software in the whole system will be hard, I'd want to make a very localized change that lets me send TraceParent to W, while leaving everything else untouched to minimize work and risk of regression. To accomplish this I might want to logically do something like this:

```
Activity oldCurrent = Activity.Current;
Activity w3cActivity = new Activity();
Activity.Current = w3cActivity.SetParentId(ConvertRequestIDToW3CID(oldCurrent.Id));
// send http to W
Activity.Current = oldCurrent; 
```

vancem commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +347,77 @@ public void Stop()
+        /// However if the activity has no parent, it has to do something.   
+        /// This sets it so that WC3 format is used in this case.  
+        /// </summary>
+        static public bool UseW3CFormat { get; set; }
```
I considered an enum as well.   It certinaly more natural if there are multiple formats.   But you can make public statics work with multiple formats too (UseXXXFormat).    Thus it becomes a question of whether you think there will be multiple formats.   I am hoping for only 1, which is why I opted for the boolean. 

But I am not against the enum.  I am OK changing it to use that.

noahfalk replied:
I think even if there is never another format the enum is still clearer from the API standpoint IMO. Writing:


`Activity.DefaultFormat = ActivityFormat.RequestID`

vs. 

`Activity.UseW3CFormat = false`

lmolkova commented:
Regarding old HttpClient + new DiagnosticSource. They do ship together, but the pace users update is different.

ApplicationInsights will enforce new DiagnosticSource, while it may take years before users update to .NET Core 3.0. 

As a tracing system, we cannot require users to upgrade to latest .NET version. what you propose is that we detect that users do not have W3C enabled ASP.NET Core & HttpClient and turn off W3C for them. 

This will de-facto block W3C adoption and will create a discrepancy between different versions. 

Note that there is no way to convert Request-Id to W3C - old root Id is not compatible with new TraceId format. 
So even on the tracing system side, we cannot support both formats without huge implications on the backend performance.

SergeyKanzhelev commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +350,88 @@ public void Stop()
+        /// it is expected to be special cased (it has its own HTTP header), it is more 
+        /// convenient/efficient if it is not lumped in with other baggage.   
+        /// </summary>
+        public string TraceState
```
So the fact it is a string, not a list is still up with discussions, correct? Should it be marked in comments that this API is not final?

SergeyKanzhelev commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +350,88 @@ public void Stop()
+        /// it is expected to be special cased (it has its own HTTP header), it is more 
+        /// convenient/efficient if it is not lumped in with other baggage.   
+        /// </summary>
+        public string TraceState
```
Spec uses tracestate as a word. Please do not upper case `State`. See discussion https://github.com/census-instrumentation/opencensus-specs/pull/171

NinoFloris replied:
Other http headers/consts that I know of in Microsoft/System.Net are written with .net naming conventions.

SergeyKanzhelev replied:
`Tracestate` follows naming convention if considered to be a single word.

SergeyKanzhelev commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +350,88 @@ public void Stop()
+        /// to store vendor-specific information that must flow.
+        /// Logically it is just a kind of baggage (if flows just like baggage), but because
+        /// it is expected to be special cased (it has its own HTTP header), it is more 
+        /// convenient/efficient if it is not lumped in with other baggage.   
```
I don't think this summary is helpful for the person uses this API. I'd suggest the following text:

> The filed `tracestate` carries information supplemental to trace identity contained in `traceparent`. List of key value pairs carries by `tracestate` convey information about request position in multiple distributed tracing graphs. It is typically used by distributed tracing systems and should not be used as a general purpose baggage as this use may break correlation of a distributed trace.

vancem replied:
I will add this to the comment

SergeyKanzhelev commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +350,88 @@ public void Stop()
+        {
+            get
+            {
+                for (var activity = this; activity != null; activity = activity.Parent)
```
Add a check for an infinite loop.

vancem replied:
After reviewing the code I believe that it is impossible for the parent chain to have a loop.  I will add a comment as to the reasoning.

SergeyKanzhelev commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +573,22 @@ private partial class KeyValueListNode
+    public enum ActivityIDFormat : byte
+    {
+        Unknown,     // ID format is not known.     
+        RequestID,   // |XXXX.XX.XX ...
```
should `RequestID` be called `HierarchicalRequestID`? `|` tells that it is hierarchical, anything else - it's flat.

SergeyKanzhelev replied:
Also place use this example in comment so it will be clear that underscores are expected: `|XXXX.XX_YY.XX_YY ...`

lmolkova commented:
> As an absolute minimum of changes to support W3C it may work. 

I do not agree. We should understand what is the default and configurable behavior for Id format. Now it's defined by upstream service.

We should allow either enforce Request-Id or W3C:  deterministic behavior controlled by current service.

SergeyKanzhelev commented:
@lmolkova I agree with you that we need to make it deterministic. This implementation make it an ASP.NET decision as parsing logic may decide to switch Activity to the new format when receiving `Request-ID`. But it will require a setting for ASP.NET which may not be desirable.

Thus I'm saying it is an OK first step that may require many more steps

vancem commented:
I have added a strawman for Activity sampling.   The idea is to make an 'always on' capability cheap enough so that we will indeed leave it always on (just turn the sampling down.  

Basically on incoming HTTP request (in RecordRequestStartEventLog) we now will do something like this
```
            Activity activity = Activity.CreateWithWithSampling(ActivityName, requestId);
            if (activity != null)
            {
                 pull out all baggage and other state from the headers).  
```   

Logging systems can basically issue config (requests) to have a certain sampling level (SamplingPercent).  Setting this to  15 will cause on average 15 out of every 100 events to be sampled.     

The activity class keeps track of all the outstanding request  for sampling and basically chooses the largest.   When the logging system disposes the config then that request for sampling goes away.

The sampling system tries to keep the given sampling rate as well as honor any explicit requests to sample (e.g. from a W3C id or from SetRecordingDesired.

SergeyKanzhelev commented:
@vancem  would it be possible to make sampling proposal a separate PR? So we can keep making progress and sampling discussing will not block possible other discussions

vancem commented:
I have pulled the activity sampling logic into its own PR https://github.com/dotnet/corefx/pull/33605.

noahfalk commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -314,7 +314,8 @@ public Activity Start()
+                if (IdFormat == ActivityIdFormat.W3C || defaultFormat == ActivityIdFormat.ForceW3C || (ParentId == null && defaultFormat == ActivityIdFormat.W3C))
                 {
                     Id = GenerateW3CId();
                     IdFormat = ActivityIdFormat.W3C;
```
It seems odd to potentially assign IdFormat = Hierarchial in SetParentId, then reverse the decision after calling Start(). That provides callers an opportunity to observe both values. I suggest that if we are forcing an id format then IdFormat property should either be uninitialized or the forced value at all times.

lmolkova commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -607,5 +608,6 @@ public enum ActivityIdFormat : byte
         Unknown,      // ID format is not known.     
         Hierarchical, //|XXXX.XX.X_X ... see https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md#id-format
         W3C,          // 00-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX-XXXXXXXXXXXXXXXX-XX see https://w3c.github.io/trace-context/
+        ForceW3C,     // This value is only used by DefaultIdFormat.   It indicate the W3C format bit it also indicate W3C should be preferred over the parent ID format.  
```
> Scenario level - Assume I own service A which calls out to service B. I want to use W3C in A, but I can't control B and it uses Hierarchical. Is the expectation that I should be able to have a trace that covers both with some type of interop in the middle? If yes, the most obvious way of doing that interop is creating an Activity with a hierarchial ID format before calling into the http client code, but I won't be able to do that if ForceW3C is set.

To achieve back-compat we have to emit both: W3C and hierarchical in corresponding headers. This will satisfy all possible scenarios. We've discussed it a bit and my proposal is to inject Request-Id with hierarchical Id on ApplicaitonInsights level - we'll gradually deprecate it.

SergeyKanzhelev commented:
@karelz this is active PR. Perhaps re-opening to clean up commits is a good idea =). @vance, can you please mention us if you will decide to open a separate PR

karelz commented:
Is the "NO MERGE" still needed? Do we have rough ETA for the PR?

vancem commented:
> Is the "NO MERGE" still needed? Do we have rough ETA for the PR?

Yes, the PR is still in a state that we would not want to check in.    We want this in for 3.0, so we expect to finalize it this month.

lmolkova commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +357,141 @@ public void Stop()
+        private string GenerateW3CId()
+        {
+            string newSpanId = Guid.NewGuid().ToString("n").Substring(16, 16);
+            if (ParentId != null)
```
I believe if ParentId is hierarchical, but W3C is forced, this code does not produce valid W3C traceparent.
This scenario is expected when AspNetCore 2.1 will be used with new DiagnosticSource.

vancem commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +347,77 @@ public void Stop()
+        /// Returns true if the ID happens to be in W3C format XX-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX-XXXXXXXXXXXXXXXX-XX
+        /// this is a convenience method.  
+        /// </summary>
+        public bool IsW3CFormat { get; private set; }
```
Now called IdFormat

vancem commented:
OK, I think that I am very close to being happy with this change to check in.    In particular I have submitted it for API review (see #34828) and  is close to being signed off.  The other major thing I have done recently is made it so that all the common ID operations are POSSIBLE to do without ever creating a string (you use the ActivityTraceId and ActivitySpanId types instead).   I modified everything to be lazy, so if you don't ask for strings they don't get created, so it should be possible for the common paths to be exercised without doing string allocation. 

I also added Testing of all the new APIs and fixed any bugs I found.   I am reasonably happy with it.  I hope to check it in early next week after the API review is complete.  

The changes are in theory completely additive.   All existing behavior still works by default.  You have to opt-in to all new behavior (WC3 IDs).

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+            _asHexString = null;
+            fixed (ulong* idPtr = &_id1)
+            {
+                var outBuff = new Span<byte>(idPtr, sizeof(ulong));
```
Rather than using `fixed` and pointers like this, can you instead do something like:
```C#
Span<byte> outBuff = MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref _id1, 1));
```
?

vancem replied:
The System.Diagnostics.DiagnosticsSource is an out of band DLL that can be used on all .NET Platforms (in particular .NET Desktop).   I do need things that work everywhere.    Is the above suggestion still viable given that?

stephentoub replied:
> Is the above suggestion still viable given that?

Not as written unfortunately, since MemoryMarshal.CreateSpan is core-only, and there's no OOB-friendly way to create a span that represents an existing field.

If you wanted, you could instead do:
```C#
Span<byte> outBuff = stackalloc byte[sizeof(ulong)];
… // logic that copies to outBuff
_id = MemoryMarshal.Read<ulong>(outBuff);
```

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+            StringBuilder sb = new StringBuilder(bytes.Length * 2);
+            foreach (byte b in bytes)
+                sb.Append(BinaryToHexDigit(b >> 4)).Append(BinaryToHexDigit(b));
+            return sb.ToString();
```
I agree it would be nice if we had a public helper for this, but in the meantime using StringBuilder like this seems like overkill and inefficient.  Assuming the input `bytes` is relatively small, you could instead do:
```C#
Span<char> result = stackalloc char[bytes.Length];
int pos = 0;
foreach (byte b in bytes)
{
    result[pos++] = BinaryToHexDigit(b >> 4);
    result[pos++] = BinaryToHexDigit(b);
}
return result.ToString();
```
or if it's larger than you'd want to put on the stack, get a buffer from the array pool, use it in the same way, and then return it.  That's what ValueStringBuilder would do, anyway.

vancem replied:
I agree that I don't like the StringBuilder.  It took your suggestion to use a stackalloc span.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -77,7 +87,21 @@ public partial class Activity
+                if (_parentId == null)
+                {
+                    if (_parentSpanIdSet)
+                        _parentId = "00-" + _traceId.AsHexString + "-" + _parentSpanId.AsHexString + "-00";
```
If we're going to have AsHexString create a StringBuilder anyway, it'd be nice to avoid doing that and creating each consituent string here, then the string[] to store all of these pieces for the concat, and then the final string... instead it'd be nicer to just create the StringBuilder (or ValueStringBuilder) here, pass it into AsHexString, and end up with just the resulting string allocation.  To use ValueStringBuilder, you can just include the relevant .cs file... that's what we do in other corefx projects right now, since it's not public.

vancem replied:
I do see that extra allocations here.   We can actually use the same stackalloc trick here (we actually to know it is fixed length),  However I would like to hold off because the theory is that under hot code paths we don't use the 'Id' property (we ask for SpanId and TraceIds).   Thus if we do our job right, this path will not be 'hot' and thus worthy of complexity to make it more efficient.     If we find that that in practice the 'Id' property is hot we can fix this.  

I have put a comment to that effect in the code.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +409,173 @@ public void Stop()
+        {
+            get
+            {
+                for (var activity = this; activity != null; activity = activity.Parent)
```
Nit: var => whatever type `activity` is

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +409,173 @@ public void Stop()
+            {
+                for (var activity = this; activity != null; activity = activity.Parent)
+                {
+                    var val = activity._traceState;
```
Nit: var => whatever type _traceState is

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +409,173 @@ public void Stop()
+        /// If the Activity has the W3C format, this returns the ID for the SPAN part of the Id.  
+        /// Otherwise it returns a zero SpanId. 
+        /// </summary>
+        public ref readonly ActivitySpanId SpanId
```
Interesting... why ref readonly?

vancem replied:
This is because SpanId caches things (basically its conversion to a Hex string), and lazily populates it.   Thus to get the benefit of that caching they need to be passed by reference (It is also 1 pointer rather than sizeof(Span) which is twice as big).   It is definately true that there is performance subtelty here (it would be easy to copy it out rather than keep it as a ref), but we really only expect a modest number of clients who really need to get this right.

I have put a comment in both the SpanId and TraceId doc-comment about this to make it more explicit.

stephentoub replied:
> It is definately true that there is performance subtelty here (it would be easy to copy it out rather than keep it as a ref)

In particular, won't any method calls / property accessors accessed off of the ref readonly instance result in C# making a copy on which to make the call?

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +409,173 @@ public void Stop()
+            {
+                if (!(ActivityIdFormat.Hierarchical <= value && value <= ActivityIdFormat.W3C))
+                    throw new ArgumentException($"value must be a valid ActivityIDFormat value");
+                s_DefaultIdFormat = value;
```
The resulting race conditions here are ok?

vancem replied:
I think so.   The expectation is that both DefaultIdFormat and ForceDefaultIdFormat will be set by the logging system before any interesting logging, and not change thereafter  (it is a setup value).    There is an issue that multiple logging systems could 'fight' over the property's value, but the expectation is that that was a configuration error (you should not try to have two logging system that have different ideas about what the ID format it).

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -343,7 +409,173 @@ public void Stop()
+            //  * 2 chars flags 
+            //  = 55 chars (see https://w3c.github.io/trace-context)
+            // We require that all non-WC3IDs NOT start with a digit.  
+            return id.Length == 55 && char.IsDigit(id[0]);
```
char.IsDigit checks for more than just '0'-'9'... is that ok?

vancem replied:
What does IsDigit do besides 0..9?   
Note that the goal here was a very fast way of looking at a string and separating valid Hierarchical IDs (which always being with | from W3C Ids (which always begin with 0-9).   It is better if it is as  'narrow' as possible, so if IsDigit returns true for more than 0..9 it is probably better to replace it with an explicit range check.

stephentoub replied:
> What does IsDigit do besides 0..9?

It factors in Unicode digits:
https://github.com/dotnet/coreclr/blob/99b0e1a1729d354d4444c8fbd4149df30f7f2940/src/System.Private.CoreLib/shared/System/Char.cs#L209-L216

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+                    idData.CopyTo(outBuff);
+                }
+            }
+        }
```
This looks identical to ActivitySpanId's ctor.  Can they use a shared helper?

vancem replied:
I looked at this, and yes you could make a helper but it is not clear it makes the code better.   The code in this constructor is really not doing much (an 'if check' (on whether it is UTF8, an argument check and calls helper to do the work).  Thus most of the lines of code that would be shared is just the argument error check..   If we would have had separate overloads for the UTF8 case and the binary case,  frankly this method would have looked like the ctor(Span<char> case and we would not feel like it needed a helper.   

I don't have a big care factor, so if you really want me to factor that I will, but I would prefer to just leave it.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+            fixed (ulong* idPtr = &_id1)
+            {
+                if (idData.Length != 32)
+                    throw new ArgumentOutOfRangeException(nameof(idData));
```
You can move this up to the beginning of the ctor, since it's argument validation.

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+            {
+                fixed (ulong* idPtr = &_id1)
+                    return new ReadOnlySpan<byte>(idPtr, sizeof(ulong) * 2);
+            }
```
This looks dangerous. What if this ActivityTraceId is stored on a class?  The GC could move the instance in memory, thus also moving the location of the ActivityTraceId and its _id1 field, and now this ReadOnlySpan is pointing to the wrong place.

vancem replied:
AsBytes removed replaced with CopyTo

stephentoub commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+                fixed (ulong* idPtr = &_id1)
+                    return new ReadOnlySpan<byte>(idPtr, sizeof(ulong));
+            }
+        }
```
Same concern as on ActivityTraceId.

ericstj commented:
I added a workaround for the package cycle.  This is going to fail package testing due to the dangling ref.  Let me add a workaround for that.

vancem commented:
Reminding myself that the API review issue for this new API is  https://github.com/dotnet/corefx/issues/34828

ahsonkhan commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +738,368 @@ private partial class KeyValueListNode
+                {
+                    if (idData.Length != 32)
+                        throw new ArgumentOutOfRangeException(nameof(idData));
+                    ActivityTraceId.SetSpanFromHexUtf8Chars(outBuff, idData);
```
This can be written more cleanly (and safely) by leveraging the `Utf8Parser`. It should also be faster. Something like:
```C#
if (idData.Length != 32)
    throw new ArgumentOutOfRangeException(nameof(idData));

Utf8Parser.TryParse(idData.Slice(0, 16), out _id1, out _, 'x');
Utf8Parser.TryParse(idData.Slice(16, 16), out _id2, out _, 'x');
if (BitConverter.IsLittleEndian)
{
    _id1 = BinaryPrimitives.ReverseEndianness(_id1);
    _id2 = BinaryPrimitives.ReverseEndianness(_id2);
}
```

ahsonkhan commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +738,368 @@ private partial class KeyValueListNode
+        /// Converts 'idData' which is assumed to be HEX UTF8 characters to binary
+        /// puts it in 'outBytes'
+        /// </summary>
+        internal static void SetSpanFromHexUtf8Chars(Span<byte> outBytes, ReadOnlySpan<byte> idData)
```
The UTF-8 character helpers shouldn't be necessary.

ahsonkhan commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -480,9 +738,368 @@ private partial class KeyValueListNode
+    }
 }
+
+
```
nit: remove extra lines

vancem commented on pkg/test/packageSettings/System.Diagnostics.DiagnosticSource/workaroundMissingSystemMemory.targets:
```diff
@@ -0,0 +1,11 @@
+
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <ItemGroup Condition="$(TargetFramework.StartsWith('netstandard1.')) OR $(TargetFramework.StartsWith('netcoreapp1.'))">
+    <PackageReference Include="System.Memory" Version="4.5.1" />
```
This is a fix that @joperezr suggested when @ericstj  was out.   I don't have a strong opinion, but in general isn't it better to use the oldest version that works?

ericstj replied:
This is just a temporary workaround in the package testing infra.  It's nothing that persists into the product.  You can tell folks to manually reference the latest.

vancem commented on src/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:
```diff
@@ -22,26 +22,83 @@ public static Activity Current
+        public void StopActivity(Activity activity, object args) { }
+    }
+
+
```
I will remove the space.  This package was relatively unusual in that its support went back to 4.5 and they split the reference assembly (into DiagnosticSourceActivity.cs) to be the part that is only inlcuded in later versions (they could have achieved this with ifdefs as well.   Anyway, this was too complex for auto-generation, but when I added new things (the ActivityTraceId ...) I did generate it with /GenerateReferenceSource and cut and paste.   Thus is is pretty close.

vancem commented on src/System.Diagnostics.DiagnosticSource/tests/Configurations.props:
```diff
@@ -2,11 +2,8 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
     <BuildConfigurations>
-      netstandard1.1;
```
We got failures trying to build these test configurations because of the way we build with xunit.   After some discussion with @ericstj we concluded that these configurations did not make sense (it did not increase coverage over the other configurations that are still present).   I can forward the e-mail conversation on it if you wish.

vancem commented on external/netfx/netfx.depproj:
```diff
@@ -38,6 +38,9 @@
     <PackageReference Include="$(TargetingPackNugetPackageId)">
       <Version >$(_TargetingPackVersion)</Version>
     </PackageReference>
+    <PackageReference Include="System.Memory" Condition="'$(TargetGroup)' == 'net45' OR '$(TargetGroup)' == 'net46'">
```
This is a question for @joperezr  or @ericstj.

vancem commented:
OK, unless someone vetos it, I am merging this later today.   I have incorperated @ahsonkhan feedback.    I do want to get packages to the App-Insights folks.

ahsonkhan commented:
> OK, unless someone vetos it, I am merging this later today

**Question:** Would we address API review feedback (if any) in a subsequent commit? I imagine people having opinions about the `isUtf8Chars` and overloading `ReadOnlySpan<byte>` to mean both binary and utf-8 text in this API which is something we don't have precedent for:
```C#
public ActivitySpanId(System.ReadOnlySpan<byte> idData, bool isUtf8Chars = false) { throw null; }
```

ahsonkhan commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -792,22 +794,25 @@ public ActivityTraceId(ReadOnlySpan<byte> idData, bool isUtf8Chars = false)
-                    if (idData.Length != 16)
-                        throw new ArgumentOutOfRangeException(nameof(idData));
-                    idData.CopyTo(outBuff);
+                    _id1 = BinaryPrimitives.ReverseEndianness(_id1);
```
I am not 100% confident about reversing the endianness here, mainly because I am not sure why we  get the bytes as big endian which ends up requiring this reversal. I would have thought we would want: `if (!BitConverter.IsLittleEndian) { reverse }`, i.e. reverse only for BigEndian. Just highlighting it.

vancem replied:
I believe we want it the way it currently is.   Basically the standard wants the bytes written in byte-by-byte order, which is 'standard' for network stuff.  This happens to correspond to big-endian if you try to clump it as larger integers as we are doing here.

vancem replied:
We did not get any feedback on the API review on this API.   We can change it, but the sooner the better, as it is a breaking change.   The alternative is to pick one to use the span<byte> constructor and the other is a static (or make both methods that take strings (as Spans) to be static).    If there is precedent for resolving this I am happy to follow it.

ahsonkhan commented on src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:
```diff
@@ -904,13 +908,14 @@ internal static void SetToRandomBytes(Span<byte> outBytes)
         /// </summary>
         internal static string SpanToHexString(ReadOnlySpan<byte> bytes)
         {
+            Debug.Assert(bytes.Length <= 16);   // We want it to not be very bing
```
nit: `very bing`

davidfowl commented:
Nice!
